x/incentive/spec/01_concepts.md

```diff
@@ -4,23 +4,86 @@ order: 1
+The quantity tracking a user's involvement is referred to as "source shares". And the total across all users the "total source shares". The quotient then gives their percentage involvement, eg if a user borrowed 10,000 usdx, and there is 100,000 usdx borrowed by all users, then they will get 10% of rewards.
 
 ## Efficiency
 
-Paying out rewards to every user every block would be slow and lead to long block times. Instead rewards are calculated much less frequently.
+Paying out rewards to every user every block would be slow and lead to long block times. Instead rewards are calculated lazily only when needed.
 
-Every block a global tracker adds up total rewards paid out per unit of user involvement. For example, per unit of xrpb supplied to hard, or per share in a kava/usdx swap pool. A user's specific reward can then be calculated as needed based on their current source shares.
+First, every block, the amount of rewards to be distributed in that block are divided by the total source shares to get the rewards per share. This is added to a global total (named "global indexes"). This is repeated every block such that the global indexes represents the total rewards a user should be owed per source share if they had held a deposit from when the rewards were created.
 
-Users' rewards must be updated whenever their source shares change. This happens through hooks into other modules that run before deposits/borrows/supplies etc.
+Then, if a user has deposited (say into a CDP) at the very start of the chain (and never changed their deposit), their current reward balance can be calculated at any time $t$ as
+
+$$
+rewards_t = global\_indexes_t * source\_shares_t
+$$
+
+If a user modifies their source shares (at say time $t-10$) we can still calculate their total rewards:
+
+$$
+rewards_t = \text{rewards accrued up to time t-10} + \text{rewards accrued from time t-10 to time t}
+$$
+$$
+rewards_t = global\_indexes_{t-10} * source\_shares_{t-10} + (global\_indexes_t - global\_indexes_{t-10}) * source\_shares_t
+$$
+
+This generalizes to any number of source share modifications.
+
+In code, to avoid storing the entire history of a user's source shares and global index values, rewards are calculated on every source shares change and added to a reward balance
+
+$$
+rewards_t = reward\_balance_{t -10} + (global\_indexes_t - global\_indexes_{t-10}) * source\_shares_t
+$$
+
+Old values of $reward\_balance$ and $global\_indexes$ ares stored in a `Claim` object for each user.
+
+Listeners on external modules fire to update these values when source shares change. For example when a user deposits to hard, a method in incentive is called that calculates the rewards accrued since the last time their deposit changed. These must be called whenever source shares change, otherwise incorrect rewards will be distributed.
```

Comment:
I think this could include updating the reward_indexes

Reply:
I added some more details about how the claim object is updated